### ISSUE TITLE
Stateful rule that always runs last

### ISSUE THREAD

**Issue by git-afsantos** [HypothesisWorks/hypothesis#1365]
Hello,

**TL;DR:** Is there a way to guarantee that a rule will always be selected as the last one in a trace, so as to avoid spending time on useless examples?

This is not exactly a feature request. If you can provide me a yes/no answer and a couple hints I might be able to do it myself. :slightly_smiling_face: 

---

I am using the stateful testing API to test an asynchronous publisher-subscriber interface. I understand that Hypothesis was not designed for this, and might not be the right tool for the job, but so far I have been able to achieve some promising results.

One of the things that made me use the stateful API instead of `@given` is the minimal falsifying trace that is printed at the end. But, for my use case, I would find it useful to define a rule that always runs *last* (pretty much the opposite of `@initialize`). This is because some properties depend on incoming messages that are sent either periodically, or as a response to a published message. As such, examples that just publish but do not wait for incoming messages are useless.

Illustrating example:

```python
state = Tester()
state.publish_topic1(msg=...)
state.publish_topic2(msg=...)
state.teardown()
```

A more useful example:

```python
state = Tester()
state.publish_topic1(msg=...)
state.spin() # this waits for (and processes) incoming messages
state.teardown()
```

I have been trying to monkey patch the stateful API to give me this feature. I tried to add a `@finalize` decorator, and even to patch `StateMachineRunner.run` to directly call a `finalize()` method. It kinda worked, but it is not consistent.

This is my current code.

```python
def monkey_patch_run(self, state_machine, print_steps=None):
    if print_steps is None:
        print_steps = current_verbosity() >= Verbosity.debug
    self.data.hypothesis_runner = state_machine

    should_continue = cu.many(
        self.data, min_size=1, max_size=self.n_steps,
        average_size=self.n_steps,
    )

    try:
        if print_steps:
            state_machine.print_start()
        state_machine.check_invariants()

        while should_continue.more():
            value = self.data.draw(state_machine.steps())
            if print_steps:
                state_machine.print_step(value)
            state_machine.execute_step(value)
            state_machine.check_invariants()

        if hasattr(state_machine, "finalize"):
            if print_steps:
                state_machine.print_finalize()
            state_machine.finalize()
            state_machine.check_invariants()
    finally:
        if print_steps:
            state_machine.print_end()
        state_machine.teardown()
```

The main issue, as I understand it, is that the `StateMachineRunner.run` loop uses `conjecture.utils.many()` as its iteration condition -- which makes it impossible to know beforehand when the loop is going to end, and which iteration is the last. This is why I added the call to `finalize()` after the loop.

The second issue, that still makes this inconsistent, is that some examples will not get to the end of the loop, by throwing `hypothesis.errors.StopTest`. I have not fully understood yet in which cases `StopTest` is raised, so I do not know whether it matters.

**Comment by Zac-HD** [comment]
What you want to do is (strictly speaking) not part of the stateful testing model and therefore not supported.

That said, your code looks like the right way to add a finalize method!  I'd be open to discussion of adding it to Hypothesis (cc @DRMacIver), like we recently did for `@initialize` in #1291 - which would also fix the fairly dangerous state of relying on internal interfaces...

**Comment by DRMacIver** [comment]
I'm a little confused by what you're trying to do that wouldn't be achievable by just implementing the `teardown` method. Could you elaborate?

On the `StopTest` issue: That's an internal exception, so if you're seeing that it's a Hypothesis bug! If you could provide an example of how to trigger that it would be very helpful.

**Comment by git-afsantos** [comment]
> I'm a little confused by what you're trying to do that wouldn't be achievable by just implementing the `teardown` method. Could you elaborate?

Sure, implementing the actions I want to run last in `teardown` was my initial approach. This may be nitpicking, but it felt like I was *abusing* the intent of the method. It felt cleaner to have `teardown` just perform clean up operations (which, in principle, should not raise an exception), and have something else perform some final asserts.

This also made the example trace clearer and easier to understand (as the error would not originate from `teardown`, but from other meaningful action).

I am open to implement what I want within `teardown` if there are no better alternatives, but since I was not sure, I felt like asking here. In any case, I can always override `print_end` to better illustrate what is going on.

> On the `StopTest` issue: That's an internal exception, so if you're seeing that it's a Hypothesis bug!

No issues here. I was seeing `StopTest` within my own implementation of `StateMachineRunner.run`, not in *user code*. :slightly_smiling_face:

**Comment by Zac-HD** [comment]
OK, I think our formal advice is that the supported way of doing this is some combination of not worrying about useless rules (they'll be shrunk away if there's an error), using `@teardown()` where that works, and using `@precondition(...)` to filter rules if you need efficiency.

Feel free to keep posting if there's anything more to discuss!
